From 2e6eec53b9cfdb8de4bf17fcc94cac7dcce8edb1 Mon Sep 17 00:00:00 2001
Message-Id: <2e6eec53b9cfdb8de4bf17fcc94cac7dcce8edb1.1354713829.git.Jiri.Denemark@gmail.com>
From: Martin Kletzander <mkletzan@redhat.com>
Date: Wed, 5 Dec 2012 13:50:14 +0100
Subject: [PATCH] Make deleting storage files default with added prompt

https://bugzilla.redhat.com/show_bug.cgi?id=878946

This patch changes the default checkbox-state of "Delete all
associated storage" to be checked, but adds a prompt with a warning
for users to be sure they notice this change and they know what they
are doing (hopefully).

(crobinso: confirm even if no gconf schema available, cosmetic spacing
 fixup, add Martin to AUTHORS)
(cherry picked from commit b2a7c396de511f47bedfdff172a2835b32b63a0e)

Signed-off-by: Jiri Denemark <Jiri.Denemark@gmail.com>
---
 AUTHORS                        |  2 +-
 src/virt-manager.schemas.in    | 13 +++++++++++++
 src/virtManager/config.py      | 13 ++++++++++++-
 src/virtManager/delete.py      | 15 +++++++++++++--
 src/virtManager/preferences.py | 11 ++++++++++-
 src/vmm-preferences.glade      | 29 +++++++++++++++++++++++++++++
 6 files changed, 78 insertions(+), 5 deletions(-)

diff --git a/AUTHORS b/AUTHORS
index 884bba5..4501957 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -71,8 +71,8 @@ Further patches have been submitted by:
   Gerhard Stenzel <gerhard-dot-stenzel-at-de-dot-ibm-dot-com>
   Yuri Chornoivan <yurchor-at-ukr-dot-net>
   Joey Boggs <jboggs-at-redhat-dot-com>
+  Martin Kletzander <mkletzan-at-redhat-dot-com>
   <...send a patch & get your name here...>
 
 Also lots of translations from the Fedora translation team. See
 the individual .po files for the translators.
-
diff --git a/src/virt-manager.schemas.in b/src/virt-manager.schemas.in
index dfcd7ba..e662495 100644
--- a/src/virt-manager.schemas.in
+++ b/src/virt-manager.schemas.in
@@ -378,6 +378,19 @@
     </schema>
 
     <schema>
+      <key>/schemas/apps/::PACKAGE::/confirm/delete_storage</key>
+      <applyto>/apps/::PACKAGE::/confirm/delete_storage</applyto>
+      <owner>::PACKAGE::</owner>
+      <type>bool</type>
+      <default>1</default>
+
+      <locale name="C">
+        <short>Confirm deleting storage</short>
+        <long>Whether we require a confirmation on deleting storage</long>
+      </locale>
+    </schema>
+
+    <schema>
       <key>/schemas/apps/::PACKAGE::/manager_window_height</key>
       <applyto>/apps/::PACKAGE::/manager_window_height</applyto>
       <owner>::PACKAGE::</owner>
diff --git a/src/virtManager/config.py b/src/virtManager/config.py
index 33345f5..ef81a6a 100644
--- a/src/virtManager/config.py
+++ b/src/virtManager/config.py
@@ -1,5 +1,5 @@
 #
-# Copyright (C) 2006 Red Hat, Inc.
+# Copyright (C) 2006, 2012 Red Hat, Inc.
 # Copyright (C) 2006 Daniel P. Berrange <berrange@redhat.com>
 #
 # This program is free software; you can redistribute it and/or modify
@@ -398,6 +398,13 @@ class vmmConfig(object):
         return self.conf.get_bool(self.conf_dir + "/confirm/interface_power")
     def get_confirm_unapplied(self):
         return self.conf.get_bool(self.conf_dir + "/confirm/unapplied_dev")
+    def get_confirm_delstorage(self):
+        # If no schema is installed, we _really_ want this to default to True
+        path = self.conf_dir + "/confirm/delete_storage"
+        ret = self.conf.get(path)
+        if ret == None:
+            return True
+        return self.conf.get_bool(path)
 
 
     def set_confirm_forcepoweroff(self, val):
@@ -412,6 +419,8 @@ class vmmConfig(object):
         self.conf.set_bool(self.conf_dir + "/confirm/interface_power", val)
     def set_confirm_unapplied(self, val):
         self.conf.set_bool(self.conf_dir + "/confirm/unapplied_dev", val)
+    def set_confirm_delstorage(self, val):
+        self.conf.set_bool(self.conf_dir + "/confirm/delete_storage", val)
 
     def on_confirm_forcepoweroff_changed(self, cb):
         return self.conf.notify_add(self.conf_dir + "/confirm/forcepoweroff", cb)
@@ -425,6 +434,8 @@ class vmmConfig(object):
         return self.conf.notify_add(self.conf_dir + "/confirm/interface_power", cb)
     def on_confirm_unapplied_changed(self, cb):
         return self.conf.notify_add(self.conf_dir + "/confirm/unapplied_dev", cb)
+    def on_confirm_delstorage_changed(self, cb):
+        return self.conf.notify_add(self.conf_dir + "/confirm/delete_storage", cb)
 
 
     # System tray visibility
diff --git a/src/virtManager/delete.py b/src/virtManager/delete.py
index 22f3b23..fe7da70 100644
--- a/src/virtManager/delete.py
+++ b/src/virtManager/delete.py
@@ -1,5 +1,5 @@
 #
-# Copyright (C) 2009 Red Hat, Inc.
+# Copyright (C) 2009, 2012 Red Hat, Inc.
 # Copyright (C) 2009 Cole Robinson <crobinso@redhat.com>
 #
 # This program is free software; you can redistribute it and/or modify
@@ -97,7 +97,7 @@ class vmmDeleteDialog(vmmGObjectUI):
         self.widget("delete-cancel").grab_focus()
 
         # Disable storage removal by default
-        self.widget("delete-remove-storage").set_active(False)
+        self.widget("delete-remove-storage").set_active(True)
         self.widget("delete-remove-storage").toggled()
 
         populate_storage_list(self.widget("delete-storage-list"),
@@ -126,6 +126,17 @@ class vmmDeleteDialog(vmmGObjectUI):
     def finish(self, src_ignore):
         devs = self.get_paths_to_delete()
 
+        if devs:
+            ret = util.chkbox_helper(self,
+                                     self.config.get_confirm_delstorage,
+                                     self.config.set_confirm_delstorage,
+                                     text1=_("Are you sure you want to delete "
+                                             "all the storage?"),
+                                     text2=_("This will delete all selected "
+                                             "storage data."))
+            if not ret:
+                return
+
         self.topwin.set_sensitive(False)
         self.topwin.window.set_cursor(gtk.gdk.Cursor(gtk.gdk.WATCH))
 
diff --git a/src/virtManager/preferences.py b/src/virtManager/preferences.py
index fbf2f8e..2a748af 100644
--- a/src/virtManager/preferences.py
+++ b/src/virtManager/preferences.py
@@ -1,5 +1,5 @@
 #
-# Copyright (C) 2006 Red Hat, Inc.
+# Copyright (C) 2006, 2012 Red Hat, Inc.
 # Copyright (C) 2006 Daniel P. Berrange <berrange@redhat.com>
 #
 # This program is free software; you can redistribute it and/or modify
@@ -48,6 +48,7 @@ class vmmPreferences(vmmGObjectUI):
         self.add_gconf_handle(self.config.on_confirm_removedev_changed(self.refresh_confirm_removedev))
         self.add_gconf_handle(self.config.on_confirm_interface_changed(self.refresh_confirm_interface))
         self.add_gconf_handle(self.config.on_confirm_unapplied_changed(self.refresh_confirm_unapplied))
+        self.add_gconf_handle(self.config.on_confirm_delstorage_changed(self.refresh_confirm_delstorage))
 
         self.refresh_view_system_tray()
         self.refresh_update_interval()
@@ -67,6 +68,7 @@ class vmmPreferences(vmmGObjectUI):
         self.refresh_confirm_removedev()
         self.refresh_confirm_interface()
         self.refresh_confirm_unapplied()
+        self.refresh_confirm_delstorage()
 
         self.window.signal_autoconnect({
             "on_prefs_system_tray_toggled" : self.change_view_system_tray,
@@ -87,6 +89,7 @@ class vmmPreferences(vmmGObjectUI):
             "on_prefs_confirm_removedev_toggled": self.change_confirm_removedev,
             "on_prefs_confirm_interface_toggled": self.change_confirm_interface,
             "on_prefs_confirm_unapplied_toggled": self.change_confirm_unapplied,
+            "on_prefs_confirm_delstorage_toggled": self.change_confirm_delstorage,
             "on_prefs_btn_keys_define_clicked": self.change_grab_keys,
             "on_prefs_graphics_type_changed": self.change_graphics_type,
             "on_prefs_storage_format_changed": self.change_storage_format,
@@ -219,6 +222,10 @@ class vmmPreferences(vmmGObjectUI):
                                   ignore3=None, ignore4=None):
         self.widget("prefs-confirm-unapplied").set_active(
                                 self.config.get_confirm_unapplied())
+    def refresh_confirm_delstorage(self, ignore1=None, ignore2=None,
+                                   ignore3=None, ignore4=None):
+        self.widget("prefs-confirm-delstorage").set_active(
+                                self.config.get_confirm_delstorage())
 
     def grabkeys_get_string(self, keysyms):
         keystr = None
@@ -315,6 +322,8 @@ class vmmPreferences(vmmGObjectUI):
         self.config.set_confirm_interface(src.get_active())
     def change_confirm_unapplied(self, src):
         self.config.set_confirm_unapplied(src.get_active())
+    def change_confirm_delstorage(self, src):
+        self.config.set_confirm_delstorage(src.get_active())
 
     def change_graphics_type(self, src):
         gtype = 'vnc'
diff --git a/src/vmm-preferences.glade b/src/vmm-preferences.glade
index b8ccc5f..569fac9 100644
--- a/src/vmm-preferences.glade
+++ b/src/vmm-preferences.glade
@@ -792,6 +792,35 @@ QCOW2</property>
                                 <property name="bottom_attach">6</property>
                               </packing>
                             </child>
+                            <child>
+                              <widget class="GtkLabel" id="label66">
+                                <property name="visible">True</property>
+                                <property name="can_focus">False</property>
+                                <property name="xalign">0</property>
+                                <property name="label" translatable="yes">Deleting storage:</property>
+                              </widget>
+                              <packing>
+                                <property name="top_attach">7</property>
+                                <property name="bottom_attach">8</property>
+                                <property name="x_options">GTK_FILL</property>
+                              </packing>
+                            </child>
+                            <child>
+                              <widget class="GtkCheckButton" id="prefs-confirm-delstorage">
+                                <property name="visible">True</property>
+                                <property name="can_focus">True</property>
+                                <property name="receives_default">False</property>
+                                <property name="use_action_appearance">False</property>
+                                <property name="draw_indicator">True</property>
+                                <signal name="toggled" handler="on_prefs_confirm_delstorage_toggled" swapped="no"/>
+                              </widget>
+                              <packing>
+                                <property name="left_attach">1</property>
+                                <property name="right_attach">2</property>
+                                <property name="top_attach">7</property>
+                                <property name="bottom_attach">8</property>
+                              </packing>
+                            </child>
                           </widget>
                         </child>
                       </widget>
-- 
1.8.0

